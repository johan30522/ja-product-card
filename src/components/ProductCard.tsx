import React,{CSSProperties} from 'react';
import { createContext } from 'react';
import styles from "./../styles/styles.module.css";


import { useProduct } from '../hooks/useProduct';
import { InitialValues, OnChangeArgs, Product, ProductCardHandlers, ProductContextProps } from '../interfaces/interfaces';





export interface Props {
    product: Product;
    // children?: ReactElement | ReactElement[];
    children?: (arg:ProductCardHandlers) => JSX.Element;
    className?: string;
    style?: CSSProperties;
    onChange?: (args:OnChangeArgs) => void;
    //  onChange?: () => void;
    value?: number;
    initialValues?: InitialValues
}

// const productContext=createContext({})
export const productContext = createContext<ProductContextProps>({} as ProductContextProps)

const { Provider } = productContext;


export const ProductCard = ({ children, product, className,style,onChange,value=0, initialValues }: Props) => {


    const { counter, handleAdd,isMaxCountReached,reset } = useProduct({onChange,product,value,initialValues});


    return (

        <Provider value={{
            counter,
            handleAdd,
            product,
            maxCount:initialValues?.maxCount
        }}>
            <div className={`${styles.productCard} ${className}`} style={style}>
                {
                    children && children({
                        count: counter,
                        product: product,
                        handleAdd,
                        isMaxCountReached,
                        reset  
                    })
                }
            </div>
        </Provider>
    )
}


// ProductCard.Title = ProductTitle;
// ProductCard.Img = ProductImg;
// ProductCard.Buttons = ProductButtons;


