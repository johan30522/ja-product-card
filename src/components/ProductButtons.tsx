
import React,{ useContext, CSSProperties, useCallback } from "react";
import { productContext } from "./ProductCard";

import styles from "./../styles/styles.module.css";


export interface Props {
    className?: string;
    activeBtnClass?: string;
    style?: CSSProperties;
}


export const ProductButtons = ({ className, activeBtnClass, style }: Props) => {

    const { counter, handleAdd, maxCount } = useContext(productContext);



    //callback isMaxReached , dependecies: counter,maxCount
    const isMaxReached = useCallback(() => {
        if (maxCount) {
            return counter >= maxCount;
        } else {
            return false;
        }
    }, [counter, maxCount])

    console.log('isMaxReached', isMaxReached());

    return (
        <div className={`${styles.buttonsContainer} ${className}`} style={style}>
            <button
                onClick={() => handleAdd(-1)}
                className={`${styles.buttonMinus} ${activeBtnClass}`}>-</button>
            <div className={styles.countLabel}>{counter}</div>
            <button
                disabled={isMaxReached()}
                onClick={() => handleAdd(1)}
                className={`${styles.buttonAdd} ${isMaxReached() ? styles.buttonDisabled : activeBtnClass
                    }`}>+</button>

            {/* className={`${styles.buttonAdd} ${activeBtnClass}`}>+</button> */}
        </div>
    )
}
