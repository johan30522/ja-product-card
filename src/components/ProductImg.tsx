import React,{ useContext, CSSProperties } from 'react';
import { productContext } from './ProductCard';
import styles from "./../styles/styles.module.css";


import noimage from "./../assets/no-image.jpg";

export interface Props {
    className?: string;
    image?: string;
    style?: CSSProperties;
}

export const ProductImg = ({ image = '', className, style }:Props
) => {

    const { product } = useContext(productContext);

    let imgToshow = image || product.image || noimage;



    return (

        <img
            className={`${styles.productImg} ${className}`}
            style={style}
            src={
                imgToshow
            } alt="Product" />
    )
}