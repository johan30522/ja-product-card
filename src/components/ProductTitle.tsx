import React,{ useContext,CSSProperties } from "react";
import { productContext } from "./ProductCard";
import styles from "./../styles/styles.module.css";



export interface Props {
    name?: string;
    className?: string;
    style?: CSSProperties;
}


export const ProductTitle = ({ name = '' ,className='',style }: Props) => {

    const { product } = useContext(productContext);

    let nameToShow = name || product.name;

    return (
        <span className={`${styles.productDescription} ${className}`} style={style}>{nameToShow}</span>
    )
}