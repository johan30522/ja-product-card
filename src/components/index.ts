import { ProductCard as ProductCardHOC } from './ProductCard';



import { ProductImg } from './ProductImg';
import { ProductTitle } from './ProductTitle';
import { ProductButtons } from './ProductButtons';
import { ProductCardHOCProps } from '../interfaces/interfaces';


// export * from './ProductCard';
export { ProductImg }  from './ProductImg';
export { ProductTitle } from './ProductTitle';
export { ProductButtons } from './ProductButtons';

export const ProductCard:ProductCardHOCProps = Object.assign(ProductCardHOC,{
    Img: ProductImg,
    Title: ProductTitle,
    Buttons: ProductButtons
}); 

export default ProductCard;