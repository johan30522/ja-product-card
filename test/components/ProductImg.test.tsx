import React from 'react';
import { ProductImg, ProductCard } from '../../src/components';

import renderer from 'react-test-renderer';

import { product1 } from '../data/products';


describe('test for ProductImg', () => {


    test('debe de mostrar el component de imagen', () => {

        const wrapper = renderer.create(<ProductImg image="customImage.png" />).toJSON();
        console.log(wrapper);

        expect(wrapper).toMatchSnapshot();

    });

    test('debe de mostrar el component con  la imagen en su componente', () => {

        const wrapper = renderer.create(

            <ProductCard product={product1}>

           { () =>( <ProductImg />)}


            </ProductCard>
        ).toJSON();


console.log('*******************************************************************');
        console.log('wrapper');

        expect(wrapper).toMatchSnapshot();

    });


    
});
