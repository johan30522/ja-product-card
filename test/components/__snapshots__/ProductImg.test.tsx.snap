// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for ProductImg debe de mostrar el component con  la imagen en su componente 1`] = `
<div
  className="productCard undefined"
>
  <img
    alt="Product"
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;

exports[`test for ProductImg debe de mostrar el component de imagen 1`] = `
<img
  alt="Product"
  className="productImg undefined"
  src="customImage.png"
/>
`;
